py/kubeflow/testing/cleanup_ci.py

```diff
@@ -222,6 +221,44 @@ def cleanup_service_accounts(args):
   logging.info("Unexpired emails:\n%s", "\n".join(unexpired_emails))
   logging.info("expired emails:\n%s", "\n".join(expired_emails))
 
+
+def cleanup_service_account_bindings(args):
+  credentials = GoogleCredentials.get_application_default()
+  iam = discovery.build('iam', 'v1', credentials=credentials)
+  accounts = []
+  next_page_token = None
+  while True:
+    service_accounts = iam.projects().serviceAccounts().list(
+      name='projects/' + args.project, pageToken=next_page_token).execute()
+    for a in service_accounts["accounts"]:
+      accounts.append(a["email"])
+    if not "nextPageToken" in service_accounts:
+      break
+    next_page_token = service_accounts["nextPageToken"]
+
+  resourcemanager = discovery.build('cloudresourcemanager', 'v1', credentials=credentials)
+  iamPolicy = resourcemanager.projects().getIamPolicy(resource=args.project).execute()
+  keepBindings = []
+  for binding in iamPolicy['bindings']:
```

Comment:
Does this code properly handle the case where we only delete some members for some role?

The structure of the iam policy file looks like:


```
bindings:
- members:
  - serviceAccount:593963025935@cloudbuild.gserviceaccount.com
  - serviceAccount:kubeflow-releasing@kubeflow-releasing.iam.gserviceaccount.com
  role: roles/cloudbuild.builds.builder
```
So for a given role it might be the case that we only delete some members but not others.

I think the code below will keep a binding if any member is still valid.

Why not instead just do something like

members_to_keep = []


for member in binding["members"]:
    ...
    if keep:
        members_to_keep.append(member)

binding["members"]=members_to_keep

Reply:
Good point, updated